src/libawkward/Index.cpp

```diff
@@ -146,6 +148,15 @@ namespace awkward {
         out << (int64_t)getitem_at_nowrap(i);
       }
     }
+    if(ptr_lib_ == kernel::Lib::cuda_kernels) {
+      out << "]\" offset=\"" << offset_ << "\" length=\"" << length_
+          << "\" at=\"0x" << std::hex << std::setw(12) << std::setfill('0')
+          << reinterpret_cast<ssize_t>(ptr_.get())
+          << "\" on=\"[" << kernel::get_ptr_device_num(ptr_lib(), ptr_.get())
+          << "]" << kernel::get_ptr_device_name(ptr_lib(), ptr_.get())
+          << "\" Lib=\"" << "cuda_kernels" << "\"/>" << post;
+      return out.str();
+    }
```

Comment:
This is a nice touch: now I understand why you wanted the `get_ptr_device_num` function to be visible outside of the kernels.

Since the choice of kernel changes the amount of information to be displayed (if cpu-kernels, there's nothing; if cuda-kernels, there's the device number), it should be displayed as a nested XML element, rather than XML attributes. That allows it to grow indefinitely and it makes it clear that the `on` attribute is related to the `Lib` (which shouldn't be capitalized if an attribute, anyway).

```xml
<Identities ...>
    <CUDA device_num="0" at="0x12345">
</Identities>
```

The equivalent for main memory is implicit (doesn't need to be written). This is also how Identities are included in objects, by the way:

```xml
<ListArray ...>
    <Identities ... />
</ListArray>
```

and they just don't appear if there are no identities.

Reply:
```xml
<ListOffsetArray64>
    <offsets><Index64 i="[0 3 4 4 5]" offset="0" length="5" at="0x0005013c0000">
        <Lib name="cuda" device="[0]GeForce 940MX"/>
    </Index64></offsets>
    <content><ListOffsetArray64>
        <offsets><Index64 i="[0 3 3 5 6 10]" offset="0" length="6" at="0x0005013c1000">
            <Lib name="cuda" device="[0]GeForce 940MX"/>
        </Index64></offsets>
        <content><NumpyArray format="d" shape="10" data="0 1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9" at="0x0005013c2000">
            <Lib name="cuda" device="[0]GeForce 940MX"/>
        </NumpyArray></content>
    </ListOffsetArray64></content>
</ListOffsetArray64>
```
The new function prints out arrays like this. Is this fine?

Reply:
Maybe separate the device number from the device name? Having the device name at all is a nice touch! (But device number is good thing to be able to access programmatically, at least in principle, so it ought to be separate.)

Reply:
Sure, no problem!